File: chap2/sec2.1/exampleNievergeltNonLinearFig.m
% Settings for Matlab
close all; clear;
format long

f=@(t,u) cos(t)*exp(-u);                   % RHS of the ODE problem
u0=log(2); T=2*pi;                         % initial solution, final time
N=10;                                      % number of subintervals
nSteps=100;                                % fine steps per subintervals
[tPred,uPred]=ForwardEuler(f,[0,T],u0,N);  % approximate prediction
Mn=2; width=0.2;                          % Nievergelt's method parameters
solver=@(t0,t1,u0) UForwardEuler(f,t0,t1,u0,nSteps);
[U,uTraj]=Nievergelt(solver,T,u0,N,uPred,Mn,width);  % Nievergelt method

[tFine,uFine]=ForwardEuler(f,[0,T],u0,N*nSteps);

figure(1)
plot(tFine, uFine, DisplayName='Fine')
hold on; xlabel('Time'); ylabel('Solution');
plot(tPred, uPred, 'o', DisplayName='Prediction')
plot(tPred, U, '^', DisplayName='Nievergelt')
TT=linspace(0,T,N+1);
x = uTraj(1,1,:); x=x(:);
plot(linspace(TT(1),TT(2),nSteps+1), x, 'k--', DisplayName='Trajectories')
for n=2:N
    for m=1:Mn
        x = uTraj(n,m,:); x=x(:);
        plot(linspace(TT(n),TT(n+1),nSteps+1), x, 'k--', ...
            HandleVisibility='off')
    end
end
legend(Location='northeast'), set(gca,'fontsize', 12);
saveas(figure(1), '../Figures/NievergeltExampleNonLinear_1', 'epsc')

Mn=4; width=0.4;                          % Nievergelt's method parameters
solver=@(t0,t1,u0) UForwardEuler(f,t0,t1,u0,nSteps);
[U,uTraj]=Nievergelt(solver,T,u0,N,uPred,Mn,width);  % Nievergelt method

[tFine,uFine]=ForwardEuler(f,[0,T],u0,N*nSteps);

figure(2)
plot(tFine, uFine, DisplayName='Fine')
hold on; xlabel('Time'); ylabel('Solution');
plot(tPred, uPred, 'o', DisplayName='Prediction')
plot(tPred, U, '^', DisplayName='Nievergelt')
TT=linspace(0,T,N+1);
x = uTraj(1,1,:); x=x(:);
plot(linspace(TT(1),TT(2),nSteps+1), x, 'k--', DisplayName='Trajectories')
for n=2:N
    for m=1:Mn
        x = uTraj(n,m,:); x=x(:);
        plot(linspace(TT(n),TT(n+1),nSteps+1), x, 'k--', ...
            HandleVisibility='off')
    end
end
legend(Location='northeast'), set(gca,'fontsize', 12);
saveas(figure(2), '../Figures/NievergeltExampleNonLinear_2', 'epsc')

File: chap2/sec2.1/exampleNievergeltNonLinearFig.py
import numpy as np
import matplotlib.pyplot as plt
from Nievergelt import nievergelt
from ForwardEuler import forwardEuler

# Define the RHS of the ODE problem
f = lambda t, u: np.cos(t) * np.exp(-u)
u0 = np.log(2)  # Initial solution
T = 2 * np.pi  # Final time
N = 10  # Number of subintervals
nSteps = 100  # Fine steps per subinterval

# Approximate prediction using Forward Euler
tPred, uPred = forwardEuler(f, [0, T], u0, N)

# First set of Nievergelt's method parameters
Mn = 2
width = 0.2
solver = lambda t0, t1, u0: forwardEuler(f, [t0, t1], u0, nSteps)[-1].ravel()
U, uTraj = nievergelt(solver, T, u0, N, uPred, Mn, width)

# Fine solution for comparison
tFine, uFine = forwardEuler(f, [0, T], u0, N * nSteps)

# Plot the first set of results
plt.figure(1)
plt.plot(tFine, uFine, label='Fine', zorder=3)
plt.xlabel('Time')
plt.ylabel('Solution')
plt.plot(tPred, uPred, 'o', label='Prediction', zorder=2)
plt.plot(tPred, U, '^', label='Nievergelt', zorder=2)

# Plot trajectories
TT = np.linspace(0, T, N + 1)
x = uTraj[0, 0, :]
plt.plot(np.linspace(TT[0], TT[1], nSteps + 1), x, 'k--', label='Trajectories', zorder=1)

for n in range(1, N):
    for m in range(Mn):
        x = uTraj[n, m, :]
        plt.plot(np.linspace(TT[n], TT[n + 1], nSteps + 1), x, 'k--', zorder=1, alpha=0.7)

plt.legend(loc='upper right')
plt.gca().tick_params(labelsize=12)
plt.savefig('NievergeltExampleNonLinear_1.eps', format='eps')
plt.show()

# Second set of Nievergelt's method parameters
Mn = 4
width = 0.4
U, uTraj = nievergelt(solver, T, u0, N, uPred, Mn, width)

# Fine solution for comparison
tFine, uFine = forwardEuler(f, [0, T], u0, N * nSteps)

# Plot the second set of results
plt.figure(2)
plt.plot(tFine, uFine, label='Fine', zorder=3)
plt.xlabel('Time')
plt.ylabel('Solution')
plt.plot(tPred, uPred, 'o', label='Prediction', zorder=2)
plt.plot(tPred, U, '^', label='Nievergelt', zorder=2)

# Plot trajectories
x = uTraj[0, 0, :]
plt.plot(np.linspace(TT[0], TT[1], nSteps + 1), x, 'k--', label='Trajectories', zorder=1)

for n in range(1, N):
    for m in range(Mn):
        x = uTraj[n, m, :]
        plt.plot(np.linspace(TT[n], TT[n + 1], nSteps + 1), x, 'k--', zorder=1, alpha=0.7)

plt.legend(loc='upper right')
plt.gca().tick_params(labelsize=12)
plt.savefig('NievergeltExampleNonLinear_2.eps', format='eps')
plt.show()
